cpp/src/copying/segmented_shift.cu

```diff
@@ -0,0 +1,281 @@
+                                                 bool nullable,
+                                                 size_type offset,
+                                                 device_span<size_type const> segment_offsets,
+                                                 data_type value_type,
+                                                 size_type column_size,
+                                                 rmm::cuda_stream_view stream,
+                                                 rmm::mr::device_memory_resource* mr)
+{
+  if (offset > 0) {
+    auto filter = [segment_offsets, offset] __device__(auto const& i) {
+      auto segment_bound_idx =
+        thrust::upper_bound(thrust::seq, segment_offsets.begin(), segment_offsets.end(), i) - 1;
+      return not(*segment_bound_idx <= i and i < *segment_bound_idx + offset);
+    };
+    return copy_if_else(nullable,
+                        input_pair_iterator,
+                        input_pair_iterator + column_size,
+                        fill_pair_iterator,
+                        filter,
+                        value_type,
+                        stream,
+                        mr);
+  } else {
+    auto filter = [segment_offsets, offset] __device__(auto const& i) {
+      auto segment_bound_idx =
+        thrust::upper_bound(thrust::seq, segment_offsets.begin(), segment_offsets.end(), i);
+      return not(*segment_bound_idx + offset <= i and i < *segment_bound_idx);
+    };
+    return copy_if_else(nullable,
+                        input_pair_iterator,
+                        input_pair_iterator + column_size,
+                        fill_pair_iterator,
+                        filter,
+                        value_type,
+                        stream,
+                        mr);
+  }
+}
+
+/**
+ * @brief Helper function to invoke string specialization of `copy_if_else`
+ */
+template <typename PairIterator, typename ScalarIterator>
+std::unique_ptr<column> segmented_shift_string_impl(PairIterator input_pair_iterator,
+                                                    ScalarIterator fill_pair_iterator,
+                                                    size_type offset,
+                                                    device_span<size_type const> segment_offsets,
+                                                    size_type column_size,
+                                                    rmm::cuda_stream_view stream,
+                                                    rmm::mr::device_memory_resource* mr)
+{
+  if (offset > 0) {
+    auto filter = [segment_offsets, offset] __device__(auto const& i) {
+      auto segment_bound_idx =
+        thrust::upper_bound(thrust::seq, segment_offsets.begin(), segment_offsets.end(), i) - 1;
+      return not(*segment_bound_idx <= i and i < *segment_bound_idx + offset);
+    };
+    return strings::detail::copy_if_else(input_pair_iterator,
```

Comment:
Shouldn't the top level `copy_if_else` function handle the strings specific stuff for you?